### PR TITLE
Feature/refactor client reconnection

### PR DESCRIPTION
**Pull request by philippseith**
While writing a sample for the go client, I realised that the current methods for client reconnect are impractical.
Reconnecting the client makes it necessary to create a new connection, so the client itself can only reconnect if it knows how to create the connection. This is solved by the AutoReconnect option where the client gets a factory func to create a new connection.
When the client is able to auto reconnect, it needs property to show the its three possible states: Not connected, connected and canceled. It was not possible to display this with the Context() method, so it was replaced by the new Connected() method. 
To complete the API change, the Stop() method was removed, as stopping of the client could be done by canceling the context passed to NewClient()
@andig

### CODE REVIEW

**Comment by andig** [comment]
I wasn't a fan of the context api anyway which is why I've stayed on the older commit. Using that `Closed()` channel I habe implemented reconnect very simply like this:


	// subscribe for updates and retry when connection fails
	var client signalr.Client
	if client, err = c.subscribe(ts); err == nil {
		go func(signalr.Client) {
			for {
				<-client.Closed()
				for client, err = c.subscribe(ts); err != nil; {
					c.log.ERROR.Println("connect:", err)
					time.Sleep(5 * time.Second)
				}
			}
		}(client)
	}

`subscribe` will establish the http connection, connect the client and subscribe for updates.

My 2c:
- I appreciate that the returned context is gone
- having the `conn` argument feels odd. I'd suggest replacing it with something like `WithConnection` plus `WithConnectionFactory` (still not perfect though)
- you could still get rid of that complexity by returning to the `Closed()` api that existed before where creating the connection is clients responsibility, as would be re-establishing. That also removes the need for `Connected()`.

In summary, at this time, I'd probably leave reconnecting to the client side with a clear API to detect connection loss.

**Comment by philippseith** [comment]
@andig 
- removed the conn parameter and introduced WIthConnection
- renamed AutoConnect to WithAutoConnect
- renamed Receiver to WithReceiver
- introduced baseclass for receivers
- replaced Connected channel with WaitConnected and WaitClosed blocking methods (with sync.Cond.Wait): The client methods needed to wait fro connected anyway, so I added the method to the interface. The Wait methods could be used to display the connection state in an UI. WaitClosed can be used to reconnect the client, but I do not recommend this. On reconnection you have to instantiate a new client, which leads to problems or a proxy client when the client instance is needed in different goroutines or (in case of an UI) callbacks. It is better to use WithAutoConnect for such use cases.

**Comment by andig** [comment]
Will try asap, but probably only next week…

**Comment by andig** [comment]
I'm not able to get the current commit:

```
❯ go get github.com/philippseith/signalr@1321c9849a5e75f0e5ace9a8a2189edb10881762
go: downloading github.com/philippseith/signalr v0.4.2-0.20211024093649-1321c9849a5e
# github.com/philippseith/signalr
../../go/pkg/mod/github.com/philippseith/signalr@v0.4.2-0.20211024093649-1321c9849a5e/clientoptions.go:16:10: client.connectionFactory undefined (type *client has no field or method connectionFactory)
```

**Comment by andig** [comment]
Nvm, used the wrong commit.

When using `WithAutoReconnect`- how  can I find out if the initial, first connection has worked? Should I use `WaitConnected()` with reasonable timeout or would that try forever? Use case is to see if user credentials are valid when the first connection is made.

**Comment by andig** [comment]
I just gave it a try:

```
	client, err := signalr.NewClient(context.Background(),
		signalr.WithAutoReconnect(c.connect(ts)),
		signalr.WithReceiver(c),
		signalr.Logger(easee.SignalrLogger(c.log.TRACE), false),
	)

	if err == nil {
		ctx, cancel := context.WithTimeout(context.Background(), time.Second)
		defer cancel()
		err = client.WaitConnected(ctx)
	}

	if err == nil {
		err = <-client.Send("SubscribeWithCurrentState", c.charger, true)
	}

	return c, err
}

// subscribe connects to the signalR hub
func (c *Easee) connect(ts oauth2.TokenSource) func() (signalr.Connection, error) {
	return func() (signalr.Connection, error) {
		return signalr.NewHTTPConnection(context.Background(), "https://api.easee.cloud/hubs/chargers",
			signalr.WithHTTPHeadersOption(func() (res http.Header) {
				if tok, err := ts.Token(); err == nil {
					res = http.Header{
						"Authorization": []string{fmt.Sprintf("Bearer %s", tok.AccessToken)},
					}
				}
				return res
			}),
		)
	}
}
```

In my test, `client.WaitConnected` never returned, even when the context deadline expired. Also, the actual `connect` function is never called.

Looks like there's still a bug somewhere?

**Comment by philippseith** [comment]
You are right, there is at least one bug caused by my misunderstanding of sync.Cond. I need to rethink my solution.

**Comment by andig** [comment]
Another problem I see is with `NewConnection`. It is not really clear what the meaning of context is (lifespan of the connection or of the connection attempt?). Idiomatic Go seems to say that context should be used request-scoped. 
Should also remove the `Option` suffix from the connection options as the other options outlined above do not use it (consistency).

**Comment by philippseith** [comment]
The context in NewConnection is currently scoped on the lifespan of the connection but it would be simple to extend the scope on the connection attempt.  
What is, in your opinion, the request to scope on in this case? IMHO, the user requests a connections, so the context should be used for both parts of NewConnection.

Option suffix: Same opinion, came with a PR, didn't bother yet. Will remove.

**Comment by andig** [comment]
> The context in NewConnection is currently scoped on the lifespan of the connection but it would be simple to extend the scope on the connection attempt. What is, in your opinion, the request to scope on in this case? 

As said before I'm not a big fan of using contexts for that purpose. After writing I realized that I can limit the connection attempt by passing an HTTP client. It's not super obvious though (and seem I forgot even though I wrote it...)

> IMHO, the user requests a connections, so the context should be used for both parts of NewConnection.

Using the context for the actual connection attempt seems fine, but then you can no longer use if for the lifespan of the connection.

> Option suffix: Same opinion, came with a PR, didn't bother yet. Will remove.

Thank you :)

**Comment by philippseith** [comment]
>  Also, the actual `connect` function is never called.

That's because client.Start() is missing.

**Comment by andig** [comment]
Still not working as expected:


	client, err := signalr.NewClient(context.Background(),
		signalr.WithAutoReconnect(c.connect(ts)),
		signalr.WithReceiver(c),
		signalr.Logger(easee.SignalrLogger(c.log.TRACE), false),
	)

	if err == nil {
		fmt.Println("---------- START")
		err = <-client.Start()
		fmt.Println("---------- START DONE")
	}

	if err == nil {
		ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
		defer cancel()
		fmt.Println("---------- WAIT CONNECTED")
		err = client.WaitConnected(ctx)
		fmt.Println("---------- WAIT DONE")
	}

	if err == nil {
		err = <-client.Send("SubscribeWithCurrentState", c.charger, true)
	}

The `Start` never returns, `connect` is being called now:

```
---------- START
---------- CONNECT
---------- CONNECTED
```

It is a little confusion to have `Start` return a channel of errors and have `WaitConnected` return an error as well. Would start return before it is actually connected? If not- when would one use `WaitConnected`?

**Comment by philippseith** [comment]
> It is a little confusion to have `Start` return a channel of errors and have `WaitConnected` return an error as well. Would start return before it is actually connected? If not- when would one use `WaitConnected`?

I stepped over the same problem and changed the API again. Now Start does not return an error anymore, there are State and Err properties and a method to watch for state changes. Furthermore I moved WaitConnected from the interface to an external method (WaitForClientState).
In theory, Send should not needed to be preceded with WaitForClientState, but sometimes (about  5-10 of 100 invocations) that fails. I am still investigating.

**Comment by philippseith** [comment]
The client code wasn't the main problem. ginkgos BeforeEach/AfterEach sometimes mixed instances, so tests failed sporadically. 
@andig : I hope the current solution is working now also in your case.

**Comment by philippseith** [comment]
> Another problem I see is with `NewConnection`. It is not really clear what the meaning of context is (lifespan of the connection or of the connection attempt?). Idiomatic Go seems to say that context should be used request-scoped.
> Should also remove the `Option` suffix from the connection options as the other options outlined above do not use it (consistency).

@andig I finally got the intention here :-) and changed the meaning of the context.

**Comment by andig** [comment]
That seems to work. Why is the `WaitForClientState` not part of the `Client` api though?

**Comment by philippseith** [comment]
Because `WaitForClientState`could be implemented solely by using the Client interface methods. On the other hand, it seems rather unlikely that someone will implement another client.

**Comment by andig** [comment]
...and it definitely feels strange.

**Comment by andig** [comment]
Shouldn't `WaitForClientState` be `WaitForState` for sake of simplicity now?
